crates/fuzzing/src/generators/single_inst_module.rs

```diff
@@ -0,0 +1,349 @@
+use once_cell::sync::Lazy;
+use wasm_encoder::{
+    CodeSection, ExportKind, ExportSection, Function, FunctionSection, Instruction, Module,
+    TypeSection, ValType,
+};
+
+/// The name of the function generated by this module.
+const FUNCTION_NAME: &'static str = "test";
+
+/// Configure a single instruction module.
+///
+/// By explicitly defining the parameter and result types (versus generating the
+/// module directly), we can more easily generate values of the right type.
+#[derive(Clone, Debug)]
+pub struct SingleInstModule<'a> {
+    instruction: Instruction<'a>,
+    parameters: Vec<ValType>,
+    results: Vec<ValType>,
+}
+
+impl<'a> SingleInstModule<'a> {
+    /// Generate a binary Wasm module with a single exported function, `test`,
+    /// that executes the single instruction.
+    pub fn encode(&self) -> Vec<u8> {
+        let mut module = Module::new();
+
+        // Encode the type section.
+        let mut types = TypeSection::new();
+        types.function(self.parameters.clone(), self.results.clone());
+        module.section(&types);
+
+        // Encode the function section.
+        let mut functions = FunctionSection::new();
+        let type_index = 0;
+        functions.function(type_index);
+        module.section(&functions);
+
+        // Encode the export section.
+        let mut exports = ExportSection::new();
+        exports.export(FUNCTION_NAME, ExportKind::Func, 0);
+        module.section(&exports);
+
+        // Encode the code section.
+        let mut codes = CodeSection::new();
+        let locals = vec![];
+        let mut f = Function::new(locals);
+        for (index, _) in self.parameters.iter().enumerate() {
+            f.instruction(&Instruction::LocalGet(index as u32));
+        }
+        f.instruction(&self.instruction);
+        f.instruction(&Instruction::End);
+        codes.function(&f);
+        module.section(&codes);
+
+        // Extract the encoded Wasm bytes for this module.
+        module.finish()
+    }
+}
+
+impl<'a> Arbitrary<'a> for SingleInstModule<'_> {
+    fn arbitrary(u: &mut Unstructured<'a>) -> arbitrary::Result<Self> {
+        u.choose(&INSTRUCTIONS).cloned()
+    }
+}
+
+// MACROS
+//
+// These macros make it a bit easier to define the instructions available for
+// generation. The idea is that, with these macros, we can define the list of
+// instructions compactly and allow for easier changes to the Rust code (e.g.,
+// `SingleInstModule`).
+
+macro_rules! binary {
+    ($inst:ident, i32) => {
+        binary! { $inst, i32, ValType::I32, ValType::I32 }
+    };
+    ($inst:ident, i64) => {
+        binary! { $inst, i64, ValType::I64, ValType::I64 }
+    };
+    ($inst:ident, f32) => {
+        binary! { $inst, f32, ValType::F32, ValType::F32 }
+    };
+    ($inst:ident, f64) => {
+        binary! { $inst, f64, ValType::F64, ValType::F64 }
+    };
+    ($inst:ident, $from_rust_ty:ident, $from_encoder_ty:expr,  $to_encoder_ty:expr) => {
```

Comment:
I like this set of macros!

I notice none of the `rust_ty` macro arguments are used currently. As an alternative, what about a helper macro like:

```rust
macro_rules! wasm_type {
    (i32) => { ValType::I32 };
    ...
    (f64) => { ValType::F64 };
}
```

Then use Rust types in all macro arguments, and convert them using `wasm_type!` only when constructing the `parameters` and `results` arrays. That makes the macro invocations less verbose and also means you can enhance the macros to make use of the Rust types later. You could define `convert!` and `compare!` in one line each at that point.

You could also collapse all the current macros into one like `inst!(F64Sub (f64, f64) -> (f64))`:

```rust
macro_rules! inst {
    ($inst:ident ($($param_ty:ident),*) -> ($($ret_ty:ident),*)) => {
        SingleInstModule {
            instruction: $inst,
            parameters: &[$(wasm_type!($param_ty)),*],
            results: &[$(wasm_type!($ret_ty)),*],
        }
    }
}
```

Reply:
Yeah, thanks for the comment; I cleaned things up a bit but not all the way to the final `inst!` suggestion. In the project this came from, there was more involved and so this was an exercise in "paring down" the original code. I'll leave it as it is in 28f2636 until I see a bit more clearly how this will be used in the actual fuzz target, since that might change the shape of this again.